---
title: "Figure3_PeakTiling"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("ggplot2")
library("dplyr")
library("tidyr")
require("cowplot")
library("Biostrings")
library(reshape2)
library(scales)

options(stringsAsFactors = F)
options(scipen = 10000)

data_all <- read.table('../../processed_data/peak_tile/peak_tile_expression_formatted_std.txt',
                       header = T)

negatives <- filter(data_all, category == 'neg_control')

peak_data <- data_all %>% 
    filter(category == 'tile')

```

```{r}
cor <- cor(peak_data$RNA_exp_sum_1, peak_data$RNA_exp_sum_2)

ggplot(peak_data, aes(RNA_exp_sum_1, RNA_exp_sum_2)) + 
    geom_point() +
    scale_x_log10() + scale_y_log10() + annotation_logticks(sides= 'bl') +
    annotate("text", label = paste("italic(r)==", signif(cor, 3)),
             parse = TRUE, x = 1, y = 15, size = 8) +
    xlab('Technical Replicate #1') + ylab('Technical Replicate #2') +
    ggtitle('Correlation between replicate TSS measurements') +
    geom_point(data = negatives, aes(RNA_exp_sum_1, RNA_exp_sum_2), color = "firebrick1") +
    annotate("text", label = "Negative Controls", x = 20, y = .5, 
             color = 'firebrick1', size = 8) +
    theme(axis.text.x = element_text(size = 24), 
          axis.text.y = element_text(size = 24),
          axis.title.x.bottom = element_text(size = 24),
          axis.title.y.left = element_text(size = 24))

ggsave('../../figs/fig3/peak_tile_correlation.pdf', units = 'in', 
       height = 5, width = 7)
```

Figure3B, MraZ Promoter

```{r cars}
threshold <- 1
peak_data %>% 
    filter(grepl('89077_89780', name)) %>% 
    ggplot(aes(tile_start + peak_start, expn_med_fitted_scaled)) + 
    geom_point(aes(color = active)) +
    scale_y_continuous(breaks = seq(0.5, 3, 0.5)) +
    geom_hline(yintercept = threshold, linetype = 'dashed') +
    geom_line(linetype = 'dashed') +
    scale_color_manual(values = c('red', 'black')) +
    labs(x = 'Tile Start Position', y = 'Expression', color = '', title = "mraZ Promoter") +
    annotate("text", label = "Activity Threshold", x = 89125, y = threshold+.1) +
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20),
          title = element_text(size = 24),
          legend.text = element_text(size = 20))
    

ggsave("../../figs/fig3/fig3b_MraZPromoter.pdf", units = 'in',
       height = 5, width = 7)
```

Subset active promoters and print for merging
```{r}
peak_data %>% filter(expn_med_fitted_scaled > threshold) %>% nrow()  #14912 sequences with more active expression

peak_data %>% filter(expn_med_fitted_scaled > threshold) %>%
    mutate(genome = "U00096.2",
           left = as.integer(tile_start+peak_start),
           right = tile_end+peak_start) %>% arrange(left) %>%
    select(genome, left, right, name, expn_med_fitted_scaled, strand) %>%
    write.table(., "../../processed_data/peak_tile/active_tiles.bed", 
                row.names = F, quote = F, sep = '\t', col.names = FALSE)
```


Identify overlapping active regions

```{bash}
bedtools merge -i ../../processed_data/peak_tile/active_tiles.bed -d -100 -s -c 5,6 -o median,distinct > ../../processed_data/peak_tile/merged_active_tiles.bed
```

```{r}
read.table("../../processed_data/peak_tile/merged_active_tiles.bed") %>%
    mutate(name = paste(V2, "_", V3, "_", V5, sep = "")) %>%
    select(V1, V2, V3, name, V5, V4) %>%
    write.table(., "../../processed_data/peak_tile/merged_active_tiles.bed",
                  col.names = FALSE,
                  sep = '\t',
                  row.names = F,
                  quote = F)
```

Identify active tiles that overlap peaks
```{bash}
bedtools intersect -c -a ../../processed_data/frag/lb/U00096.2_plus_minus_called_peaks_threshold1.1_merge40_min60.bed -b ../../processed_data/peak_tile/merged_active_tiles.bed -s > ../../processed_data/peak_tile/collapsed_per_peak.bed
                                    
```


3C - number of promoters per peak
```{r}
collapsed_per_Peak <- read.table("../../processed_data/peak_tile/collapsed_per_peak.bed",
                                 col.names = c("chrom", 'start', 'end', 'name', 'peak_score',
                                               'strand', 'score1', 'score2', 'num_sub_peaks')) %>% 
    separate(name, into = c('peak_start', 'peak_end', 'strand'), sep = '_', remove = F, convert = T) %>% 
    # only consider peaks with mapped tiles, eliminates 123 peaks
    semi_join(peak_data, by = c('peak_start', 'peak_end'))

collapsed_per_Peak %>% 
    filter(num_sub_peaks < 10) %>% 
    ggplot(aes(num_sub_peaks)) +
        geom_bar(stat = "count", width = .8, fill = '#528ECB', color = 'black') +
        geom_text(stat='count', aes(label=..count..), vjust=-.2, size = 10) +
        scale_y_log10() + annotation_logticks(sides = 'l') + 
        labs(x = 'Number of promoter regions per peak', y = "Count") + 
        scale_x_continuous(breaks = pretty_breaks(n=7)) +
         theme(axis.text = element_text(size = 20),
              axis.title = element_text(size = 20),
              legend.text = element_text(size = 20))

ggsave("../../figs/fig3/promoters_per_peak.pdf", units = 'in',
       height = 8, width = 10)

```


3D) Identify minimal promoter unit, 
```{r}

minimal_promoters <- read.table("../../processed_data/peak_tile/merged_active_tiles.bed", header = F,
                                col.names = c("genome", 'left', 'right', 'name', 'score', 'strand')) %>% 
                                mutate(left_minimal = right - 150,
                                right_minimal = left + 150,
                                minimal_length = right_minimal - left_minimal) 

# sizes of minimal promoters
ggplot(filter(minimal_promoters, minimal_length > 0), aes(minimal_length)) + 
      geom_histogram(color = 'black',
                     binwidth = 10,
                     fill = '#528ECB',
                     color = 'black') +
      scale_x_continuous(breaks = pretty_breaks(n=8)) + 
      labs(x = 'Length of minimum promoter region',
           y = 'Count',
           title = 'Sizes of minimum promoter regions') +
     theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20),
          legend.text = element_text(size = 20))

ggsave("../../figs/fig3/fig3d_minimal_promoter_size.pdf", units = 'in', 
       width = 7, height = 5)
```


Get fasta sequence of merged active tiles
```{bash}
bedtools getfasta -fi ../../ref/Escherichia_coli_K-12_MG1655.fasta -bed ../../processed_data/peak_tile/merged_active_tiles.bed -s -tab -fo ../../processed_data/peak_tile/merged_active_tiles.fasta
```

Figure 5E classification predictions

```{r}
roc_info <- read.table('../../processed_data/peak_tile/20190423_peak_tile_classifier_hyperparam_tuned_roc_info.txt',
                       sep = ',', col.names = c('fpr', 'tpr', 'threshold'))

auc <- pracma::trapz(roc_info$fpr, roc_info$tpr)

ggplot(roc_info, aes(fpr, tpr)) + geom_line() +
    labs(x = 'false positive rate', y = 'true positive rate') +
    annotate('text', x = 0.85, y = 0.10, label = paste('AUC = ', signif(auc, 3))) +
    theme(aspect.ratio = 1)
```

```{r}
pr_info <- read.table('../../processed_data/peak_tile/20190423_peak_tile_classifier_hyperparam_tuned_pr_info.txt',
                      sep = ',', col.names = c('precision', 'recall', 'threshold'))


auprc <- pracma::trapz(pr_info$precision, pr_info$recall)

ggplot(pr_info, aes(precision, recall)) + geom_line() +
    scale_x_continuous(breaks = c(0.00, 0.25, 0.50, 0.75, 1.00)) +
    annotate('text', x = 0.90, y = 0.15, label = 'AUC = 0.194') +
    theme(aspect.ratio = 1)
```

```{r}
num_positive <- 1516
num_negative <- 2902
tmp <- left_join(pr_info, roc_info, by = 'threshold') %>% 
    mutate(num_tp = tpr * num_positive,
           num_fp = fpr * num_negative,
           precision_new = num_tp / (num_tp + num_fp))
pr_info_new <- filter(tmp, !is.na(precision_new))  
auprc <- pracma::trapz(pr_info_new$precision_new, pr_info_new$recall)
```

Figure 5F Graph regression predictions

```{r}
reg_predictions <- read.table('../../processed_data/peak_tile/20190422_peak_tile_hyperparam_tuned_predictions.txt',
                              col.names = c('sequence', 'predicted', 'observed')) %>% 
    left_join(select(data_all, sequence = variant, expn_med_fitted_scaled, active), by = 'sequence') %>% 
    distinct(sequence, observed, .keep_all = T)

active_predictions <- filter(reg_predictions, active == 'active')
corr_active <- cor.test(active_predictions$observed, active_predictions$predicted)$estimate
inactive_predictions <- filter(reg_predictions, active == 'inactive')
corr_inactive <- cor.test(inactive_predictions$observed, inactive_predictions$predicted)$estimate
```

```{r}
reg_predictions %>% 
    ggplot(aes(observed, predicted)) + geom_point(alpha = 0.50, aes(color = active)) +
    geom_smooth(method = 'lm', aes(group = active), color = 'black') +
    scale_color_manual(values = c('navyblue', 'gray38')) +
    scale_x_log10(breaks = c(0, 0.01, 0.1, 1, 10, 100)) + 
    scale_y_log10(breaks = c(0, 0.01, 0.1, 1, 10, 100)) +
    annotation_logticks() + 
    annotate('text', x = 45, y = 0.05, label = paste0('r = ', signif(corr_inactive, 1))) +
    annotate('text', x = 45, y = 0.10, color = 'navyblue',
             label = paste0('r = ', signif(corr_active, 3))) +
    labs(x = 'observed expression', y = 'predicted expression',
         title = '',
         color = '') +
    theme(aspect.ratio = 1, legend.position = 'top')

ggsave('../../figs/fig3/peak_tile_regression.pdf', units = 'in', 
       width = 8, height = 5)
```



