---
title: "Figure3_PeakTiling"
output: html_document
---

```{r setup, include=FALSE}

library(dplyr)
library(ggplot2)
library(tidyr)
require(cowplot)
library(sqldf)
library(reshape2)
library(Biostrings)
library(tidyverse)
library(scales)

options(stringsAsFactors = F)
options(scipen = 10000)

data_all <- read.table('../../processed_data/peak_tile/peak_tile_expression_formatted_std.txt',
                       header = T)

negatives <- filter(data_all, category == 'neg_control')

peak_data <- data_all %>% 
    filter(category == 'tile')

```


Figure3B, MraZ Promoter

```{r}
threshold <- 1
peak_data %>% 
    filter(grepl('89077_89780', name)) %>% 
    ggplot(aes(tile_start + peak_start, expn_med_fitted_scaled)) + 
    geom_point(aes(color = active)) +
    scale_y_continuous(breaks = seq(0.5, 3, 0.5)) +
    geom_hline(yintercept = threshold, linetype = 'dashed') +
    geom_line(linetype = 'dashed') +
    scale_color_manual(values = c('red', 'black')) +
    labs(x = 'Tile Start Position', y = 'Expression', color = '', title = "mraZ Promoter") +
    annotate("text", label = "Threshold", x = 89125, y = threshold+.1) +
    theme(axis.text = element_text(size = 20),
          axis.title = element_text(size = 20),
          title = element_text(size = 24),
          legend.text = element_text(size = 20))

ggsave("../../figs/fig3/fig3b_MraZPromoter.pdf", units = 'in',
       height = 5, width = 7)
```

Subset active promoters and print for merging
```{r}
peak_data %>% 
    filter(expn_med_fitted_scaled > threshold) %>% 
    nrow()  #15013 sequences with more active expression

peak_data %>% 
    filter(expn_med_fitted_scaled > threshold) %>%
    mutate(genome = "U00096.2",
           left = as.integer(tile_start+peak_start),
           right = tile_end+peak_start) %>% arrange(left) %>%
    select(genome, left, right, name, expn_med_fitted_scaled, strand) %>%
    write.table(., "../../processed_data/peak_tile/active_tiles.bed", 
                row.names = F, quote = F, sep = '\t', col.names = FALSE)
```


Identify overlapping active regions

```{bash}
bedtools merge -i ../../processed_data/peak_tile/active_tiles.bed -d -100 -s -c 5,6 -o median,distinct > ../../processed_data/peak_tile/merged_active_tiles.bed
```

```{r}
read.table("../../processed_data/peak_tile/merged_active_tiles.bed") %>%
    mutate(name = paste(V2, "_", V3, "_", V4, sep = "")) %>%
    select(V1, V2, V3, name, V5, V4) %>%
    write.table(., "../../processed_data/peak_tile/merged_active_tiles.bed",
                  col.names = FALSE,
                  sep = '\t',
                  row.names = F,
                  quote = F)
```

Identify active tiles that overlap peaks
```{bash}
bedtools intersect -c -a ../../processed_data/frag/lb/U00096.2_plus_minus_called_peaks_threshold1.1_merge40_min60.bed -b ../../processed_data/peak_tile/merged_active_tiles.bed -s > ../../processed_data/peak_tile/collapsed_per_peak.bed
                                    
```

3C - number of promoters per peak
```{r}
collapsed_per_Peak <- read.table("../../processed_data/peak_tile/collapsed_per_peak.bed",
                                 col.names = c("chrom", 'start', 'end', 'name', 'peak_score',
                                               'strand', 'score1', 'score2', 'num_sub_peaks')) %>% 
    separate(name, into = c('peak_start', 'peak_end', 'strand'), sep = '_', remove = F, convert = T) %>% 
    # only consider peaks with mapped tiles, eliminates 123 peaks
    semi_join(peak_data, by = c('peak_start', 'peak_end'))

collapsed_per_Peak %>% 
    filter(num_sub_peaks < 10 & num_sub_peaks > 0) %>% 
    ggplot(aes(num_sub_peaks)) +
        geom_bar(stat = "count", width = .8, fill = '#528ECB', color = 'black') +
        geom_text(stat = 'count', aes(label = ..count..), vjust = -0.2, size = 10) +
        scale_y_log10() + annotation_logticks(sides = 'l') + 
        labs(x = 'Number of promoter\nregions per peak', y = "Count") + 
        scale_x_continuous(breaks = pretty_breaks(n = 7)) +
        theme(axis.text = element_text(size = 20),
              axis.title = element_text(size = 20),
              legend.text = element_text(size = 20))

ggsave("../../figs/fig3/promoters_per_peak.pdf", units = 'in',
       height = 4, width = 5)

#How many promoters do we identify?

sum(collapsed_per_Peak$num_sub_peaks) #2,228

```


3D) Identify minimal promoter unit, 
```{r}
minimal_promoters <- read.table("../../processed_data/peak_tile/merged_active_tiles.bed", header = F,
                                col.names = c("genome", 'left', 'right', 'name', 'score', 'strand')) %>% 
                                mutate(left_minimal = right - 150,
                                right_minimal = left + 150,
                                minimal_length = right_minimal - left_minimal,
                                name = paste(left_minimal, "_", right_minimal, "_", strand, sep = '')) %>%
                                filter(right_minimal - left_minimal > 0)

minimal_promoters %>% 
    select(genome, left_minimal, right_minimal, name,score, strand) %>%
    write.table(., "../../processed_data/peak_tile/minimal_promoters.bed", 
                row.names = F, quote = F, col.names = FALSE, sep = '\t')

# sizes of minimal promoters
a <- ggplot(filter(minimal_promoters, minimal_length > 0), aes(minimal_length)) + 
    geom_histogram(color = 'black',
                   binwidth = 10,
                   fill = '#528ECB',
                   color = 'black') +
    scale_x_continuous(breaks = pretty_breaks(n=8)) + 
    labs(x = 'Length of minimum promoter',
         y = 'Count',
         title = 'Sizes of minimum promoter regions') +
    geom_point(shape = 25, fill = 'firebrick3', color = 'navyblue', 
               aes(x = 40, y = 300), size = 3) +
    geom_point(shape = 25, fill = 'navyblue', color = 'navyblue', 
               aes(x = 150, y = 450), size = 3) +
    theme(axis.text = element_text(size = 20),
          axis.text.x = element_text(size = 20, angle = 45, hjust = 1),
          axis.title = element_text(size = 20),
          legend.text = element_text(size = 20))

b <- minimal_promoters %>% 
    filter(minimal_length > 35 & minimal_length < 45) %>%
    ggplot(aes(score)) + 
    geom_density(color = 'firebrick3', alpha = 0.4) +
    scale_x_log10(limits = c(NA, 30)) + 
    labs(y = 'Density', x = '')

c <- minimal_promoters %>% 
    filter(minimal_length == 150) %>%
    ggplot(aes(score)) + 
    geom_density(color = 'navyblue', alpha = 0.4) +
    scale_x_log10(limit = c(NA,30)) + 
    labs(y = 'Density', x = 'Minimal Promoter Expression')

dens <- plot_grid(b,c, ncol = 1)
plot_grid(a,dens, ncol = 2, rel_heights = c(1, .8))


ggsave("../../figs/fig3/fig3d_minimal_promoter_size.pdf", units = 'in', 
       width = 7, height = 5)
```


Get fasta sequence of merged active tiles
```{}
bedtools getfasta -fi ../../ref/Escherichia_coli_K-12_MG1655.fasta -bed ../../processed_data/peak_tile/merged_active_tiles.bed -s -tab -fo ../../processed_data/peak_tile/merged_active_tiles.fasta
```

Identify intragenic minimal promoters and create control intragenic regions
```{bash}
# minimal promoters overlapping genes
bedtools intersect  -b ../../ref/U00096.2_genes_clean.bed -a  ../../processed_data/peak_tile/minimal_promoters.bed > ../../processed_data/peak_tile/intragenic_promoter_gene_overlap.bed

#generate random sequences with overlap
#9393928
#7127372
#bedtools shuffle -seed 9393928 -incl ../../ref/U00096.2_genes_clean.bed -excl ../../processed_data/peak_tile/intragenic_promoter_gene_overlap.bed -i ../../processed_data/peak_tile/intragenic_promoter_gene_overlap.bed -g ../../ref/U00096.2.genome > ../../processed_data/peak_tile/intragenic_control_gene_overlap.bed

```


Get genes and translate
```{r}
genes_fasta <- read.table("../../ref/regulondb_genes.txt", skip = 8, fill = NA, sep = '\t') %>%
    subset(., !grepl("ins", .$V2)) %>%
    select( V1, gene = V2, left = V3, right = V4, strand = V5, sequence = V6) %>%
    mutate(strand = gsub("forward", "+", strand),
           strand = gsub("reverse", "-", strand),
           sequence = as.character(sequence),
           left = as.numeric(left)
           ) %>%
    na.omit() 

genes_fasta <- genes_fasta[1:6] 

genes_fasta$aa <- DNAStringSet(genes_fasta$sequence) %>% 
    Biostrings::translate() %>% 
    as.character()
```

Match genes and overlapping intragenic promoters/controls
```{r}
intragenic_promoters <- read.table("../../processed_data/peak_tile/intragenic_promoter_gene_overlap.bed",
                                   header = F,
                                   col.names = c("genome", 'prom_left', 'prom_right', 'name', 'score', 'prom_strand')) %>%
                         mutate(name = paste(name, "(", prom_strand, ")", sep = ''),
                                prom_left = as.numeric(prom_left) + 1, 
                                prom_right = as.numeric(prom_right))


# Match promoters with overlapping genes
promoter_gene_overlap_codon <- sqldf("SELECT a.*, b.*
                   FROM genes_fasta AS a
                   LEFT JOIN intragenic_promoters AS b
                   ON b.prom_left BETWEEN a.left AND a.right
                   OR b.prom_right BETWEEN a.left AND a.right") %>%
                   na.omit() %>% distinct() %>%
                   mutate(prom_start = ifelse(strand == '+', prom_left - left, right - prom_right),
                          prom_end = ifelse(strand == '+', prom_right - left, right - prom_left),
                          codon_start = (floor(prom_start / 3) * 3),
                          codon_end = (ceiling(prom_end / 3) * 3),
                          codon_length = (codon_end - codon_start) / 3)# %>% filter(prom_start >= 0)


```

Count codons
```{r message=FALSE, echo=FALSE}

for (i in 1:nrow(promoter_gene_overlap_codon)) {
  
    print(i)
    seq <- print(promoter_gene_overlap_codon$sequence[i])
    start <- print(promoter_gene_overlap_codon$codon_start[i])
    end <- print(promoter_gene_overlap_codon$codon_end[i])
     
    prom <- substr(seq, start + 1, end) 
    len <- nchar(prom)
    promoter_gene_overlap_codon$codon_seq[i] <- prom
    promoter_gene_overlap_codon$aa_len[i] <- len / 3
}

promoter_gene_overlap_codon <- filter(promoter_gene_overlap_codon, round(aa_len) == aa_len)

codon_stats <- promoter_gene_overlap_codon$codon_seq %>%
    paste(., collapse = "") %>%
    DNAStringSet() %>%
    oligonucleotideFrequency(width = 3, step = 3) %>%
    t() %>%
    as.data.frame() %>%
    mutate(codon = row.names(.)) %>%
    mutate(promoter_counts = V1, 
           promoter_frequency = V1 / sum(V1)) %>%
    select(-V1)

# Add genome counts of whole genome
temp <- genes_fasta %>%
    mutate(len_seq = nchar(sequence) / 3, 
           len_aa = nchar(aa)) %>%
    filter(len_seq == len_aa) 

codon_stats <- temp$sequence %>%
    paste(., collapse ="") %>%
    DNAStringSet() %>%
    oligonucleotideFrequency(width = 3, step = 3) %>%
    t() %>% 
    as.data.frame() %>%
    mutate(control_counts = V1, 
           control_frequency = V1 / sum(V1)) %>%
    select(-V1) %>% 
    cbind(codon_stats, .)

codon_stats$amino_acid <- DNAStringSet(codon_stats$codon) %>% 
    Biostrings::translate() %>% 
    as.character() 
codon_stats$amino_acid[codon_stats$codon == 'CTG'] <- 'L' #MUST DO THIS SINCE NO.INIT.CODON DOESN'T WORK,
codon_stats$amino_acid[codon_stats$codon == 'TTG'] <- 'L'

```


ChiSquare of AA frequencies
```{r}
chisq_aa <- function(aa, df) {
  print(aa)
  total_prom <- sum(df$aa_count_prom)
  total_control <- sum(df$aa_count_ctrl)
  codon_count_promoter <- df$aa_count_prom[df$amino_acid == aa]
  codon_count_control <- df$aa_count_ctrl[df$amino_acid == aa]
  count_table <- rbind(c(codon_count_promoter, codon_count_control),
                        c(total_prom, total_control))
  result <- chisq.test(count_table)
  return(result$p.value)
}

aa_table <- codon_stats %>% 
    group_by(amino_acid) %>%
    mutate(aa_count_prom = sum(promoter_counts),
           aa_count_ctrl = sum(control_counts)) %>% 
    ungroup() %>%
    select(amino_acid, aa_count_prom, aa_count_ctrl) %>% 
    distinct()

aa_Xsq <- data.frame(matrix(nrow = 21, ncol = 2))

for (i in 1:nrow(aa_table)) {
  aa <- aa_table$amino_acid[i]
  aa_Xsq$amino_acid[i] <- aa 
  aa_Xsq$p.val[i] <- chisq_aa(aa, aa_table) 
}

aa_Xsq <- aa_Xsq %>% 
    select(amino_acid, p.val) %>%
    mutate(p.val.adjust = p.adjust(p.val))
```



```{r}
signif_aa <- filter(aa_Xsq, p.val.adjust < .05)
  
sum(codon_stats$promoter_counts)
sum(codon_stats$control_counts)

temp <- codon_stats %>% 
    group_by(amino_acid) %>%
    mutate(aa_count_prom = sum(promoter_frequency), 
           aa_count_ctrl = sum(control_frequency),
           promoter_enrichment = log2(aa_count_prom / aa_count_ctrl),
           Type = ifelse(2^promoter_enrichment > 1, 'Enriched', 'Depleted')) %>%
    ungroup() %>%
    select(amino_acid, promoter_enrichment, Type, aa_count_ctrl) %>%
    distinct() %>% 
    left_join(., signif_aa) 


ggplot(temp, aes(reorder(amino_acid, promoter_enrichment), promoter_enrichment)) + 
    geom_bar(aes(fill = log2(aa_count_ctrl)), stat = "identity", color = 'black', alpha = 0.8) +
    scale_fill_gradientn(colours = c('red', 'white', 'blue')) +
    labs(x = 'Amino Acid', 
         y = 'log2(Relative AA composition)', 
         fill = 'log2(Amino Acid Frequency)') +
    theme(axis.text.x = element_text(size = 24), 
          axis.text.y = element_text(size = 24),
          axis.title.x.bottom = element_text(size = 24),
          axis.title.y.left = element_text(size = 24),
          legend.key = element_rect(color = "black")) + 
    coord_flip() +
    geom_point(data = temp[complete.cases(temp),], 
               aes(x= amino_acid,
                   y = promoter_enrichment + .05 * sign(promoter_enrichment)), shape = 8, size = 2) 


ggsave('../../figs/fig3/AA_bias.pdf', units = 'in', width = 10.29, height = 8.51)
```

Chi square test of significant codons within each group
```{r}
chisq_codon <- function(codon, df) {
  print(codon)
  total_prom <- sum(df$promoter_counts)
  total_control <- sum(df$control_counts)
  codon_count_promoter <- df$promoter_counts[df$codon == codon]
  codon_count_control <- df$control_counts[df$codon == codon]
  count_table <- rbind(c(codon_count_promoter, codon_count_control),
                       c(total_prom, total_control))
  result <- chisq.test(count_table)
  return(result$p.value)
}

codon_table_full <- data.frame(matrix(nrow = 64, ncol = 2)) %>%
                      select('codon'='X1', 'p.val'='X2') 
  
for (i in 1:nrow(aa_table)) {

    aa <- aa_table$amino_acid[i]
    print(aa)
    
    temp <- filter(codon_stats, amino_acid == aa)
    
    codon_table <- data.frame(matrix(nrow = 6))
    
    for (i in 1:nrow(temp)) {
        cd <- temp$codon[i]
        codon_table$codon[i] <- cd
        codon_table$p.val[i] <- chisq_codon(cd, temp)
    
}

codon_table_full <- distinct(codon_table) %>%
                      select(codon, p.val) %>%
                      rbind(codon_table_full, .)
}
    
codon_table_full <- codon_table_full[complete.cases(codon_table_full),] %>%
                      mutate(p.val.adjust = p.adjust(p.val, method = 'fdr'))


```


Identify codon bias between groups
```{r}
signif_codon <- filter(codon_table_full, p.val.adjust < .05)

temp <- codon_stats %>% 
    group_by(amino_acid) %>% 
    mutate(codon_frequency_prom = promoter_counts / sum(promoter_counts),
           codon_frequency_ctrl = control_counts / sum(control_counts),
           enrichment = log2(codon_frequency_prom / codon_frequency_ctrl),
           Type = ifelse(2 ^ enrichment > 1, 'Abundant', 'Depleted')) %>% 
    ungroup() %>%
    select(codon, amino_acid, enrichment, Type, codon_frequency_ctrl) %>% 
    # filter(amino_acid == 'L') %>% #Remove this for all amino acids
    distinct() %>% 
    left_join(., signif_codon, by = 'codon')

ggplot(temp,aes(codon, enrichment)) + 
    geom_bar(aes(fill = log2(codon_frequency_ctrl)), stat = "identity", 
             color = 'black', alpha = 0.8) +
    scale_fill_gradientn(colours = c('red', 'white', 'blue')) +
    labs(x = 'Codon', 
         y = 'log2(Relative Codon composition)',
         color = '', 
         title = "Differential Codon usage at Intragenic Promoters") +
        geom_hline(yintercept = 0) +
        theme(axis.text.x = element_text(size = 18, angle = 90), 
        axis.text.y = element_text(size = 18),
        axis.title.x.bottom = element_text(size = 18),
        axis.title.y.left = element_text(size = 18)) + 
        geom_point(data = temp[complete.cases(temp) ,],
                   aes(x = codon, y = enrichment + 0.05 * sign(enrichment)),
                   shape = 8, size = 2) +
        facet_grid(~amino_acid, scale = "free_x",space = 'free_x')


ggsave('../../figs/fig3/Codon_bias_genome.pdf', units = 'in', width = 21.58, height = 5.02)

```
