{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# E. coli MPRA visualization #\n",
    "\n",
    "In this notebook we will define all necessary code to run an interactive Bokeh visualization of the data presented in our paper. This is a direct counterpart to the script `ecoli_mpra_bokeh.py` and is displayed here for easier visualization. The interactive widget is displayed on the bottom and will show if you run the notebook locally, however it will not display on Github. I've added a screenshot into the notebook for convenience. \n",
    "\n",
    "If you wish to run this notebook and directly interact with the widgets, you must have all the necessary packages downloaded and run the notebook locally. You can also run the python script with `bokeh serve ecoli_mpra_bokeh.py`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bokeh.plotting import figure, output_notebook, show, ColumnDataSource, output_file\n",
    "from bokeh.models.plots import Plot\n",
    "from bokeh.models import Legend, LegendItem\n",
    "from bokeh.models.tools import HoverTool, BoxSelectTool, BoxZoomTool, PanTool, WheelZoomTool, SaveTool, ResetTool\n",
    "from bokeh.models.tickers import FixedTicker\n",
    "from bokeh.models.widgets import CheckboxGroup, RangeSlider, Tabs, TextInput, Button, RadioGroup, RadioButtonGroup, Select, Paragraph, Div\n",
    "from bokeh.layouts import column, row, WidgetBox, Spacer, gridplot\n",
    "from bokeh.models import Panel, Range1d, LinearAxis\n",
    "from bokeh.io import show, curdoc, export_svgs\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from math import pi\n",
    "from copy import deepcopy\n",
    "import sys\n",
    "\n",
    "global frag_plus_pileup\n",
    "global frag_minus_pileup\n",
    "global colors\n",
    "global genes\n",
    "global rna_plus\n",
    "global rna_minus\n",
    "\n",
    "NEW_MIN = 1\n",
    "NEW_MAX = 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1001\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1001\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error() {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < css_urls.length; i++) {\n",
       "      var url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error;\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1001\");\n",
       "  if (element == null) {\n",
       "    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.2.0.min.js\"];\n",
       "  var css_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.2.0.min.css\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.2.0.min.css\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.2.0.min.css\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {} // ensure no trailing comma for IE\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1001\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error() {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (var i = 0; i < css_urls.length; i++) {\n      var url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };var element = document.getElementById(\"1001\");\n  if (element == null) {\n    console.error(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n    return false;\n  }\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.2.0.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.2.0.min.js\"];\n  var css_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.2.0.min.css\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.2.0.min.css\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.2.0.min.css\"];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# display bokeh inline\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def norm_pileup(pileup):\n",
    "    \n",
    "    '''\n",
    "    This function normalizes the fragment expression data. It excludes all data below median expression for easier\n",
    "    visualization. However, for graphing we need a continuous line for all positions. We will set all data \n",
    "    below median to the median value, and then scale data so minimum value is 1. On log-scale, log(1) = 0 \n",
    "    so any data <= median will be set to zero on plot\n",
    "    '''\n",
    "    \n",
    "    median = np.median(pileup.expression)\n",
    "    norm_pileup = deepcopy(pileup)\n",
    "    norm_pileup.expression = np.where(pileup.expression <= median,\n",
    "                                     median, pileup.expression)\n",
    "    scalar = 1.0/median\n",
    "    norm_pileup.expression = norm_pileup.expression * scalar\n",
    "    # make sure everything is equal to 1, due to decimal multiplication some may be 0.9999\n",
    "    norm_pileup.expression = np.where(norm_pileup.expression < 1,\n",
    "                                     1, norm_pileup.expression)\n",
    "    return norm_pileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def min_max_scale(expression, new_min, new_max, max_value=None, median_norm=None):\n",
    "    \n",
    "    '''\n",
    "    Min-max normalization\n",
    "    '''\n",
    "    \n",
    "    if max_value:\n",
    "        # cap value at max so the scaling won't be skewed\n",
    "        expression = np.where(expression > max_value, max_value, expression)\n",
    "\n",
    "    if median_norm:\n",
    "        # subtract median\n",
    "        expression = expression - median_norm\n",
    "        # if negative, set to 0\n",
    "        expression = np.where(expression < 0, 0, expression)\n",
    "\n",
    "    # first min-max normalize to [0, 1]\n",
    "    min_exp = min(expression)\n",
    "    max_exp = max(expression)\n",
    "\n",
    "    norm = (expression - min_exp) / float(max_exp - min_exp)\n",
    "    norm_scaled = (norm * (new_max - new_min)) + new_min\n",
    "\n",
    "    return norm_scaled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_pileup_dataset(conditions, start, end, colors, log_transform=True):\n",
    "    \n",
    "    '''\n",
    "    Construct the fragment pileup visualization based on the given genomic coordinates and color\n",
    "    according to strand.\n",
    "    '''\n",
    "    \n",
    "    # subset relevant condition, conditions must be list\n",
    "    plus_pileup_subset = frag_plus_pileup_norm[(frag_plus_pileup_norm.condition.isin(conditions)) &\n",
    "                                         (frag_plus_pileup_norm.position.isin(range(start,end)))]\n",
    "    \n",
    "    minus_pileup_subset = frag_minus_pileup_norm[(frag_minus_pileup_norm.condition.isin(conditions)) &\n",
    "                                           (frag_minus_pileup_norm.position.isin(range(start,end)))]\n",
    "    \n",
    "    plus_pileup_subset = plus_pileup_subset.merge(colors[colors.strand == '+'], on='condition', how='left')\n",
    "    minus_pileup_subset = minus_pileup_subset.merge(colors[colors.strand == '-'], on='condition', how='left')\n",
    "    \n",
    "    # patch accepts list of lists, make one list for each combination of strand and condition\n",
    "    # x-axis position will be same for every condition and strand, define coordinates for edges at the top (pileup line) and the \n",
    "    position_unique = plus_pileup_subset.sort_values('position').position.unique()\n",
    "    position = np.hstack((position_unique, position_unique[::-1]))\n",
    "    # bottom edge along zero (positions in reverse to create closed shape)\n",
    "    bottom_edge = np.zeros(len(position))\n",
    "    # take negative of reverse coverage to display on graph\n",
    "    counts = []\n",
    "    position_all = []\n",
    "    condition_colors = []\n",
    "    for condition in conditions:\n",
    "        # plus strand\n",
    "        position_all.append(position)\n",
    "        if log_transform:\n",
    "            counts.append(np.hstack((np.log10(plus_pileup_subset.expression[plus_pileup_subset.condition == condition]),\n",
    "                               bottom_edge)))\n",
    "        else:\n",
    "            counts.append(np.hstack((plus_pileup_subset.expression[plus_pileup_subset.condition == condition],\n",
    "                                   bottom_edge)))\n",
    "        condition_colors.append(colors.color[(colors.condition == condition) & (colors.strand == '+')].to_string(index=False))\n",
    "        \n",
    "        # minus strand\n",
    "        position_all.append(position)\n",
    "        if log_transform:\n",
    "            counts.append(np.hstack((-np.log10(minus_pileup_subset.expression[minus_pileup_subset.condition == condition]),\n",
    "                               bottom_edge)))\n",
    "        else:\n",
    "            counts.append(np.hstack((-minus_pileup_subset.expression[minus_pileup_subset.condition == condition],\n",
    "                                   bottom_edge)))\n",
    "        condition_colors.append(colors.color[(colors.condition == condition) & (colors.strand == '-')].to_string(index=False))\n",
    "    \n",
    "    patches_src = pd.DataFrame({'position' : position_all, \n",
    "                               'count' : counts,\n",
    "                               'color' : condition_colors})    \n",
    "    \n",
    "    return ColumnDataSource(patches_src)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_region_genes(genes, start, end):\n",
    "    \n",
    "    '''\n",
    "    Grab genes within coordinates, with buffer of 100bp on either side\n",
    "    '''\n",
    "\n",
    "    start_buffer = start - 100\n",
    "    end_buffer = end + 100\n",
    "    region_genes = genes[(genes.start >= start_buffer) & (genes.end <= end_buffer)]\n",
    "\n",
    "    # center is midpoint of gene\n",
    "    gene_center = region_genes.start + abs(region_genes.end - region_genes.start)/2.0\n",
    "    gene_center = gene_center.tolist()\n",
    "    gene_color = '#e2b306'\n",
    "    gene_width = region_genes.end - region_genes.start\n",
    "    gene_width = gene_width.tolist()\n",
    "\n",
    "    src_gene = ColumnDataSource(data=dict(\n",
    "    gene_center = gene_center,\n",
    "    # y-center is on x-axis, i.e y = 0\n",
    "    gene_center_y = np.zeros(len(region_genes)),\n",
    "    gene_color = [gene_color] * len(region_genes),\n",
    "    gene_width = gene_width,\n",
    "    # set center of triangle to start or end depending on strand\n",
    "    tri_x = [region_genes.end.iloc[i] if region_genes.strand.iloc[i] == '+' else region_genes.start.iloc[i] \\\n",
    "           for i in range(len(region_genes))],\n",
    "    angle = [-90 if strand == '+' else 90 for strand in region_genes.strand],\n",
    "    gene_name = region_genes.name.tolist()))\n",
    "    \n",
    "    return src_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_tss_arrow(tss, start, end, threshold=1, active_color='#2e6eb7', inactive_color='grey', width=8, \n",
    "                   arrow_length=50):\n",
    "\n",
    "    '''\n",
    "    Draw TSS arrows, which consist of a vertical line at TSS position, a horizontal line at the top of the vertical\n",
    "    line, and a triangle. Color TSS by active/inactive\n",
    "    '''\n",
    "    \n",
    "    # grab TSS within start and end\n",
    "    tss_region = tss[tss.tss_position.isin(range(start, end))]\n",
    "    \n",
    "    # list of lists. Need four quad coordinates: top, bottom, left, right\n",
    "    # top and bottom are y-coordinates, left and right are x-coordinates\n",
    "    \n",
    "    # draw vertical rectangle to indicate TSS position. Above axis indicates positive strand TSS\n",
    "    # and below axis indicates negative strand. We'll keep it as a list of lists [top, bottom, left, right]\n",
    "    # and separate into different lists in ColumnDataSource\n",
    "    quad_coords = []\n",
    "    \n",
    "    # draw horizontal segment to connect to arrow, [x0, y0, x1, y1]\n",
    "    seg_coords = []\n",
    "    \n",
    "    # center of triangle is segment endpoint. Set direction of triangle based on strand\n",
    "    arrow_angle = []\n",
    "    \n",
    "    # assign color based on TSS expression above or below threshold\n",
    "    tss_color = []\n",
    "    \n",
    "    for i in range(len(tss_region)):\n",
    "        \n",
    "        expn = tss_region.RNA_exp_sum_ave.iloc[i]\n",
    "        position = tss_region.tss_position.iloc[i]\n",
    "        \n",
    "        if expn < threshold:\n",
    "            tss_color.append(inactive_color)   \n",
    "        else:\n",
    "            tss_color.append(active_color)\n",
    "        \n",
    "        if tss_region.strand.iloc[i] == '+':\n",
    "            # top, bottom, left, right\n",
    "            quad_coords.append([expn, 0, position, position + width])\n",
    "            # draw line to the right: x0, y0, x1, y1\n",
    "            seg_coords.append([position, expn, position + arrow_length, expn])\n",
    "            arrow_angle.append(-90)\n",
    "        elif tss_region.strand.iloc[i] == '-':\n",
    "            quad_coords.append([0, -expn, position, position + width])\n",
    "            # draw line to the left\n",
    "            seg_coords.append([position, -expn, position - arrow_length, -expn])\n",
    "            arrow_angle.append(90)\n",
    "        else:\n",
    "            raise Exception('Invalid strand')\n",
    "    \n",
    "    src_tss = ColumnDataSource(data=dict(\n",
    "    top = [quad_coords[i][0] for i in range(len(quad_coords))],\n",
    "    bottom = [quad_coords[i][1] for i in range(len(quad_coords))],\n",
    "    left = [quad_coords[i][2] for i in range(len(quad_coords))],\n",
    "    right = [quad_coords[i][3] for i in range(len(quad_coords))],\n",
    "    x0 = [seg_coords[i][0] for i in range(len(seg_coords))],\n",
    "    y0 = [seg_coords[i][1] for i in range(len(seg_coords))],\n",
    "    x1 = [seg_coords[i][2] for i in range(len(seg_coords))],\n",
    "    y1 = [seg_coords[i][3] for i in range(len(seg_coords))],\n",
    "    angle = arrow_angle,\n",
    "    color = tss_color))\n",
    "    \n",
    "    return src_tss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_rna_line(start, end, conditions):\n",
    "    \n",
    "    '''\n",
    "    Draw line representing RNA-seq coverage of given region and conditions\n",
    "    '''\n",
    "\n",
    "    # subset for relevant condition and position\n",
    "    rna_plus_subset = rna_plus[(rna_plus.condition.isin(conditions)) &\n",
    "                                           (rna_plus.position.isin(range(start, end)))]\n",
    "    rna_minus_subset = rna_minus[(rna_minus.condition.isin(conditions)) &\n",
    "                                           (rna_minus.position.isin(range(start, end)))]\n",
    "    rna_minus_subset.head()\n",
    "\n",
    "    # convert expression to log and flip minus expression to negative\n",
    "    src_rna = ColumnDataSource(data=dict(\n",
    "    x_plus = rna_plus_subset.position.tolist(),\n",
    "    y_plus = np.log10(rna_plus_subset.expression.tolist()),\n",
    "    x_minus = rna_minus_subset.position.tolist(),\n",
    "    y_minus = -(np.log10(rna_minus_subset.expression.tolist()))))\n",
    "\n",
    "    return src_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_region_plot(src, src_gene, src_tss, src_rna):\n",
    "    \n",
    "    '''\n",
    "    Construct pileup plot based on src information from multiple sources.\n",
    "    '''\n",
    "\n",
    "    # flatten list of lists in count column of src, find max value of absolute expression\n",
    "    frag_count_range = max(map(abs, [x for count in src.data['count'] for x in count]))\n",
    "    if len(src_rna.data['y_plus']) > 0:\n",
    "        rna_count_range = max(max(src_rna.data['y_plus']), max(abs(src_rna.data['y_minus'])))\n",
    "    else: \n",
    "        rna_count_range = 0\n",
    "\n",
    "    count_range = max(frag_count_range, rna_count_range)\n",
    "    \n",
    "    # draw blank figure of correct size with tools\n",
    "    p = figure(y_range=(-(count_range+1), count_range+1), plot_width=900, plot_height=700, \n",
    "               tools=[BoxSelectTool(), BoxZoomTool(), PanTool(), WheelZoomTool(), \n",
    "                      SaveTool(), ResetTool()],\n",
    "               toolbar_location='above')\n",
    "\n",
    "    legends = []\n",
    "    \n",
    "    # format axis and colors\n",
    "    p.xaxis.axis_label = 'position'\n",
    "    p.xaxis.major_label_orientation = pi/4\n",
    "    p.xaxis[0].formatter.use_scientific = False\n",
    "    # p.xaxis[0].ticker=FixedTicker(ticks=range(start, end, 100))\n",
    "    p.yaxis.axis_label = 'log normalized activity'\n",
    "        \n",
    "    patches = p.patches(source=src, xs='position', ys='count', \n",
    "        fill_color='color', line_color=None, alpha=0.50)\n",
    "    legends.append(LegendItem(label='promoter activity (plus strand)', \n",
    "        renderers=[patches], index=0))\n",
    "    legends.append(LegendItem(label='promoter activity (minus strand)', \n",
    "        renderers=[patches], index=1))\n",
    "    \n",
    "\n",
    "    # draw RNA lines\n",
    "    if len(src_rna.data['x_minus']) > 0:\n",
    "        plus_line = p.line(x='x_plus', y='y_plus', line_color='#528ecb', \n",
    "            line_width=2, source=src_rna)\n",
    "        legends.append(LegendItem(label='RNA-seq (plus strand)',\n",
    "            renderers=[plus_line], index=0))\n",
    "        minus_line = p.line(x='x_minus', y='y_minus', line_color='#ef8137',\n",
    "            line_width=2, source=src_rna)\n",
    "        legends.append(LegendItem(label='RNA-seq (minus strand)',\n",
    "            renderers=[minus_line], index=1))\n",
    "\n",
    "\n",
    "    # add second y-axis for TSS strength\n",
    "    max_tss = max(map(abs, src_tss.data['y0']))\n",
    "    p.extra_y_ranges = {'tss' : Range1d(start=-(max_tss * 4), end=(max_tss * 4))}\n",
    "    p.add_layout(LinearAxis(y_range_name='tss', axis_label='TSS expression'), 'right')\n",
    "    \n",
    "    # draw vertical rectangle\n",
    "    p.quad(top='top', bottom='bottom', left='left', right='right', color='color',\n",
    "           source=src_tss, y_range_name='tss')\n",
    "    # draw horizontal line for arrow\n",
    "    p.segment(x0='x0', y0='y0', x1='x1', y1='y1', color='color',\n",
    "              source=src_tss, line_width=4,  y_range_name='tss')\n",
    "    # center of triangle is endpoint of segment\n",
    "    tri = p.triangle(x='x1', y='y1', size=9, angle='angle', angle_units='deg',\n",
    "        color='color', source=src_tss, y_range_name='tss')\n",
    "    legends.append(LegendItem(label='inactive TSS', renderers=[tri], index=9))\n",
    "    legends.append(LegendItem(label='active TSS', renderers=[tri], index=0))\n",
    "    \n",
    "\n",
    "    # plot genes\n",
    "    p.rect(x='gene_center', y='gene_center_y', width='gene_width', color='gene_color', \n",
    "        height=10, height_units='screen', source=src_gene)\n",
    "    p.triangle(x='tri_x', y=0, size=20, angle='angle', angle_units='deg',\n",
    "                     fill_color='gene_color', line_color=None, source=src_gene)\n",
    "    p.text(x='gene_center', y='gene_center_y', text='gene_name', text_color='black',\n",
    "          text_align='center', text_baseline='middle', text_font_size='10pt', source=src_gene)\n",
    "    \n",
    "\n",
    "    p.add_layout(Legend(items=legends), 'below')\n",
    "\n",
    "    return p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def toggle_genes(attr, old, new):\n",
    "    \n",
    "    '''\n",
    "    Define button activity to toggle gene visualization\n",
    "    '''\n",
    "    \n",
    "    if gene_button.value == 'show': # show genes\n",
    "        # grab genes at current input position, update\n",
    "        new_src_gene = make_region_genes(genes, int(start.value), int(end.value))\n",
    "        src_gene.data.update(new_src_gene.data)\n",
    "    if gene_button.value == 'hide': # hide genes\n",
    "        # no genes\n",
    "        new_src_gene = make_region_genes(genes, 0, 0)\n",
    "        src_gene.data.update(new_src_gene.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def toggle_tss(attr, old, new):\n",
    "    \n",
    "    '''\n",
    "    Define button activity to toggle TSS arrows\n",
    "    '''\n",
    "    \n",
    "    if tss_button.value == 'show':\n",
    "        new_src_tss = make_tss_arrow(endo_tss_lb, int(start.value), int(end.value))\n",
    "        src_tss.data.update(new_src_tss.data)\n",
    "    if tss_button.value == 'hide':\n",
    "        new_src_tss = make_tss_arrow(endo_tss_lb, 0, 0)\n",
    "        src_tss.data.update(new_src_tss.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def toggle_rna(attr, old, new):\n",
    "\n",
    "    '''\n",
    "    Define button activity to toggle RNA-seq data\n",
    "    '''\n",
    "    \n",
    "    if rna_button.value == 'show':\n",
    "        new_src_rna = make_rna_line(int(start.value), int(end.value), \n",
    "            [condition_selection.labels[i] for i in condition_selection.active])\n",
    "        src_rna.data.update(new_src_rna.data)\n",
    "    if rna_button.value == 'hide':\n",
    "        new_src_rna = make_rna_line(0, 0, [condition_selection.labels[i] for i in condition_selection.active])\n",
    "        src_rna.data.update(new_src_rna.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_axis():\n",
    "    \n",
    "    '''\n",
    "    Update axis ranges to reflect current boundaries, which may have changed based on user inputs \n",
    "    since initialization\n",
    "    '''\n",
    "    \n",
    "    # flatten list of lists in count column of src, find max value of absolute expression\n",
    "    frag_count_range = max(map(abs, [x for count in src.data['count'] for x in count]))\n",
    "    if len(src_rna.data['y_plus']) > 0:\n",
    "        rna_count_range = max(max(src_rna.data['y_plus']), max(abs(src_rna.data['y_minus'])))\n",
    "    else: \n",
    "        rna_count_range = 0\n",
    "\n",
    "    count_range = max(frag_count_range, rna_count_range)\n",
    "\n",
    "    p.y_range.start = -(count_range + 1)\n",
    "    p.y_range.end = count_range + 1\n",
    "\n",
    "    # update second y-axis\n",
    "    max_tss = max(map(abs, src_tss.data['y0']))\n",
    "    p.extra_y_ranges['tss'].start = max_tss * -4\n",
    "    p.extra_y_ranges['tss'].end = max_tss * 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_bp(n=500):\n",
    "\n",
    "    '''\n",
    "    Define button activity to add 500bp to both sides of plot\n",
    "    '''\n",
    "    \n",
    "    start.value = str(int(start.value) - n)\n",
    "    end.value = str(int(end.value) + n)\n",
    "    # changing the values will trigger update_region_plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_gene(attr, old, new):\n",
    "\n",
    "    '''\n",
    "    Update plot to show gene name input to search box\n",
    "    '''\n",
    "    \n",
    "    gene_name = gene_search.value\n",
    "    gene_match = genes[genes.name == gene_name]\n",
    "    if len(gene_match != 0):\n",
    "        gene_start = gene_match.start.iloc[0]\n",
    "        gene_end = gene_match.end.iloc[0]\n",
    "        # changing start and end will automatically trigger update_region_plot\n",
    "        start.value = str(gene_start - 1500)\n",
    "        end.value = str(gene_end + 1500)\n",
    "\n",
    "    else:\n",
    "        gene_search.value = 'No gene match'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def export_svg():\n",
    "    p.output_backend ='svg'\n",
    "    export_svgs(p, filename='bokeh_plot.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "############################ Read in data ######################################\n",
    "\n",
    "# endo TSS expression\n",
    "endo_tss_lb = pd.read_csv('../processed_data/endo_tss/lb/rLP5_Endo2_lb_expression_formatted_std.txt',\n",
    "                           sep='\\t')\n",
    "\n",
    "# LB genomic shearing fragment pileup\n",
    "frag_lb_plus = pd.read_csv('../processed_data/frag/lb/plus_frag_pileup.wig',\n",
    "                            sep='\\t', skiprows=1, names=['position', 'expression'])\n",
    "\n",
    "frag_lb_minus = pd.read_csv('../processed_data/frag/lb/minus_frag_pileup.wig',\n",
    "                            sep='\\t', skiprows=1, names=['position', 'expression'])\n",
    "\n",
    "# M9 minimal genomic shearing fragment pileup\n",
    "frag_m9_plus = pd.read_csv('../processed_data/frag/m9/plus_frag_pileup_M9.wig',\n",
    "                            sep='\\t', skiprows=1, names=['position', 'expression'])\n",
    "\n",
    "frag_m9_minus = pd.read_csv('../processed_data/frag/m9/minus_frag_pileup_M9.wig',\n",
    "                            sep='\\t', skiprows=1, names=['position', 'expression'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# format frag pileup data\n",
    "frag_plus_pileup = frag_lb_plus.merge(frag_m9_plus, on='position', how='outer', suffixes=['_lb', '_m9'])\n",
    "frag_plus_pileup.columns = ['position', 'LB', 'M9']\n",
    "frag_minus_pileup = frag_lb_minus.merge(frag_m9_minus, on='position', how='outer', suffixes=['_lb', '_m9'])\n",
    "frag_minus_pileup.columns = ['position', 'LB', 'M9']\n",
    "\n",
    "frag_plus_pileup = pd.melt(frag_plus_pileup, id_vars=['position'], value_vars=['LB', 'M9'],\n",
    "             var_name='condition', value_name='expression')\n",
    "frag_minus_pileup = pd.melt(frag_minus_pileup, id_vars=['position'], value_vars=['LB', 'M9'],\n",
    "             var_name='condition', value_name='expression')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalize and scale\n",
    "frag_plus_pileup_norm = deepcopy(frag_plus_pileup)\n",
    "frag_plus_pileup_norm.expression = min_max_scale(frag_plus_pileup_norm.expression, NEW_MIN, NEW_MAX,\n",
    "    median_norm = np.median(frag_plus_pileup_norm.expression))\n",
    "frag_minus_pileup_norm = deepcopy(frag_minus_pileup)\n",
    "frag_minus_pileup_norm.expression = min_max_scale(frag_minus_pileup_norm.expression, NEW_MIN, NEW_MAX,\n",
    "    median_norm = np.median(frag_minus_pileup_norm.expression))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    3.195523e+06\n",
       "mean     2.537179e+03\n",
       "std      3.448642e+04\n",
       "min      1.000000e+00\n",
       "25%      2.000000e+00\n",
       "50%      5.000000e+00\n",
       "75%      2.600000e+01\n",
       "max      1.980450e+06\n",
       "Name: raw_expression, dtype: float64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in M9 sequencing data\n",
    "rna_plus = pd.read_csv('B6_M9_2.Forward.wig',\n",
    "    sep='\\t', skiprows=1, names=['position', 'raw_expression'])\n",
    "rna_minus = pd.read_csv('B6_M9_2.Reverse.pos_values.wig',\n",
    "    sep='\\t', skiprows=1, names=['position', 'raw_expression'])\n",
    "rna_plus.raw_expression.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalize and scale\n",
    "max_val = 1000\n",
    "rna_plus = rna_plus.assign(expression = min_max_scale(rna_plus.raw_expression, NEW_MIN, NEW_MAX, max_value=max_val,\n",
    "    median_norm=np.median(rna_plus.raw_expression)))\n",
    "rna_minus = rna_minus.assign(expression = min_max_scale(rna_minus.raw_expression, NEW_MIN, NEW_MAX, max_value=max_val),\n",
    "    median_norm=np.median(rna_minus.raw_expression))\n",
    "\n",
    "# add condition\n",
    "rna_plus['condition'] = ['M9'] * len(rna_plus)\n",
    "rna_minus['condition'] = ['M9'] * len(rna_minus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in gene annotation\n",
    "genes = pd.read_csv('U00096.2_genes_clean.bed', sep = '\\t', header = None,\n",
    "                      names=['chrom', 'start', 'end', 'name', 'score', 'strand', \n",
    "                            'thick_start', 'thick_end', 'item_rgb', 'block_count',\n",
    "                            'block_sizes', 'block_start'])\n",
    "# drop unnecessary columns and simplify name\n",
    "genes = genes[['start', 'end', 'name', 'strand']]\n",
    "genes.name = [x.split(':')[1] for x in genes.name.tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create colors for each condition and strand\n",
    "colors = pd.DataFrame({'condition' : ['LB', 'LB', 'M9', 'M9'],\n",
    "                      'strand' : ['+', '-', '+', '-'],\n",
    "                      'color' : ['#8daec9', '#edac80', '#528ecb', '#ef8137']})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def modify_doc(doc):\n",
    "    \n",
    "    def update_region_plot(attr, old, new):\n",
    "        '''\n",
    "        Update plot based on interactive widgets\n",
    "        '''\n",
    "        \n",
    "        # get list of conditions for graph\n",
    "        conditions_to_plot = [condition_selection.labels[i] for i in condition_selection.active]\n",
    "\n",
    "        position_start = int(start.value)\n",
    "        position_end = int(end.valuse)\n",
    "\n",
    "        # make new subset based on selected conditions\n",
    "        new_src = make_pileup_dataset(conditions_to_plot, position_start, position_end, colors)\n",
    "        new_src_gene = make_region_genes(genes, position_start, position_end)\n",
    "        new_src_tss = make_tss_arrow(endo_tss_lb, position_start, position_end)\n",
    "        new_src_rna = make_rna_line(position_start, position_end, conditions_to_plot)\n",
    "\n",
    "        # update source used in patch glyphs\n",
    "        src.data.update(new_src.data)\n",
    "        # update source used in gene rectangle and triangle glyphs\n",
    "        src_gene.data.update(new_src_gene.data)\n",
    "        # update source for TSS arrows\n",
    "        src_tss.data.update(new_src_tss.data)\n",
    "        # update source for RNA\n",
    "        src_rna.data.update(new_src_rna.data)\n",
    "\n",
    "    \n",
    "    ############################ Define widgets ####################################\n",
    "\n",
    "    condition_selection = CheckboxGroup(labels=['LB', 'M9'], active = [1])\n",
    "    # link change in selected buttons to update function\n",
    "    condition_selection.on_change('active', update_region_plot)\n",
    "\n",
    "    # TextInput to define start and end position\n",
    "    start = TextInput(value='1232000', title='genome (U00096.2) start position')\n",
    "    end = TextInput(value='1238000', title='genome end position')\n",
    "\n",
    "    # update plot when value is changed\n",
    "    # position_selection.on_change('value', update_region_plot)\n",
    "    start.on_change('value', update_region_plot)\n",
    "    end.on_change('value', update_region_plot)\n",
    "\n",
    "    # toggle gene annotation on and off\n",
    "    gene_button = Select(\n",
    "        title='RegulonDB gene annotation (U00096.2)',\n",
    "        options=['show', 'hide'],\n",
    "        value='show')\n",
    "    gene_button.on_change('value', toggle_genes)\n",
    "\n",
    "    # toggle TSS\n",
    "    tss_button = Select(\n",
    "        title='Transcription Start Site (TSS) condition',\n",
    "        options=['show', 'hide'],\n",
    "        value='show')\n",
    "    tss_button.on_change('value', toggle_tss)\n",
    "\n",
    "    # toggle RNA\n",
    "    rna_button = Select(\n",
    "        title='RNA-seq (M9)',\n",
    "        options=['show', 'hide'],\n",
    "        value='show')\n",
    "    rna_button.on_change('value', toggle_rna)\n",
    "\n",
    "    # add 500bp to both sides\n",
    "    add_button = Button(label='Add 500bp to both ends', button_type='primary')\n",
    "    add_button.on_click(add_bp)\n",
    "\n",
    "    # update axis range on click\n",
    "    axis_button = Button(label='Update Axis', button_type='primary')\n",
    "    axis_button.on_click(update_axis)\n",
    "\n",
    "    # search by gene name\n",
    "    gene_search = TextInput(value='Enter your favorite gene!', title='Search by gene name (case sensitive)')\n",
    "    gene_search.on_change('value', update_gene)\n",
    "\n",
    "    # export to SVG\n",
    "    export_button = Button(label='Export plot to SVG', button_type='primary')\n",
    "    export_button.on_click(export_svg)\n",
    "    \n",
    "    ############################## Initialize ######################################\n",
    "\n",
    "    # find initial conditions and position\n",
    "    initial_conditions = [condition_selection.labels[i] for i in condition_selection.active]\n",
    "    \n",
    "    initial_start = int(start.value)\n",
    "    initial_end = int(end.value)\n",
    "\n",
    "    src = make_pileup_dataset(initial_conditions,\n",
    "        initial_start,\n",
    "        initial_end,\n",
    "        colors)\n",
    "\n",
    "    src_gene = make_region_genes(genes, initial_start, initial_end)\n",
    "    src_tss = make_tss_arrow(endo_tss_lb, initial_start, initial_end)\n",
    "    src_rna = make_rna_line(initial_start, initial_end, initial_conditions)\n",
    "    \n",
    "    p = make_region_plot(src, src_gene, src_tss, src_rna)\n",
    "    \n",
    "    ################################# Layout ######################################\n",
    "\n",
    "    # combine all elements onto one page by defining layout\n",
    "\n",
    "    # add \"figure\" text\n",
    "    general_text = Div(text='''\n",
    "        <b>General:</b> The data shown here is detailed in the paper \"Genome-wide function characterization of\n",
    "        E. coli promoters\". In brief, we developed a massively parallel reporter assay that allows us to clone a\n",
    "        library of DNA sequences upstream of a GFP reporter, which in turn drives \n",
    "        expression of a 20bp barcode. We can use next-generation sequencing to quantitatively measure activity of the\n",
    "        entire library at the same time.''')\n",
    "\n",
    "    norm_text = Div(text='''\n",
    "        <b>Normalization:</b> There are several normalization steps for visualization purposes to display \n",
    "        two different datasets (promoter activity and RNA-seq) on the same axis on the left hand side. \n",
    "        First, we remove values below the median by first subtracting the median from all values, and setting negative values to zero. \n",
    "        Next, we normalize and scale the data between [1, 1000]. Finally, we log-transform the data to \n",
    "        highlight small differences in activity. ''')\n",
    "\n",
    "    frag_text = Div(text='''\n",
    "        <b>Promoter activity:</b> We created a library by randomly shearing the genome into small fragments (50-500bp) and cloned it\n",
    "        into our MPRA, measuring the promoter activity of each fragment. We tested this \n",
    "        library in two different conditions (LB and M9). We condensed the data to show average activity\n",
    "        at each nucleotide.''')\n",
    "\n",
    "    rna_text = Div(text='''\n",
    "        <b>RNA-sequencing:</b> We performed genome-wide sequencing of E. coli in M9.''')\n",
    "\n",
    "    tss_text = Div(text='''\n",
    "        <b> TSS library:</b> We synthesized a library of ~18K reported TSSs and measured their promoter activity.\n",
    "        We synthesized 150bp surrounding the TSS (-120 to +30) in LB. We classified TSSs into active\n",
    "        and inactive based on a set of 500 negative controls (150bp genomic sequence > 200bp from TSS). We set the active\n",
    "        threshold at 3 standard deviations greater than the median negative control.\n",
    "        ''')\n",
    "\n",
    "    controls_group1 = WidgetBox(general_text, norm_text, frag_text, condition_selection, rna_text, rna_button, axis_button)\n",
    "    controls_group2 = WidgetBox(gene_button, start, end, add_button, gene_search, export_button)\n",
    "    layout = row(controls_group1, p, controls_group2)\n",
    "\n",
    "    # add layout to current document\n",
    "    doc.add_root(layout)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.bokehjs_exec.v0+json": "",
      "text/html": [
       "\n",
       "<script src=\"http://localhost:56892/autoload.js?bokeh-autoload-element=1002&bokeh-absolute-url=http://localhost:56892&resources=none\" id=\"1002\"></script>"
      ]
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "server_id": "6e770172de164afa9e3046f7219f7fd9"
      }
     },
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BokehUserWarning: ColumnDataSource's columns must be of the same length. Current lengths: ('x_minus', 4480), ('x_plus', 3884), ('y_minus', 4480), ('y_plus', 3884)\n"
     ]
    }
   ],
   "source": [
    "show(modify_doc, notebook_url='http://localhost:8888')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![alt_text](./bokeh_interactive_left.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![alt_text](./bokeh_interactive_right.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda2]",
   "language": "python",
   "name": "conda-env-anaconda2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
